### PR TITLE
Featare/notifications

### PR DESCRIPTION
**Pull request by IdanPelled**
### Notification page:

The notification page binds all the messages and the invitations the user received.
to create a new message use `from app.routers.notification import create_message`.
\* bonus: `create_message` can get an extra parameter (`link`), that turns the message into a link.

#### Possible useses:
- Notify the user about a change in the meeting
- System updates

image

### CODE REVIEW

File: app/database/models.py [hunk 1]
```diff
@@ -179,6 +179,22 @@ def __repr__(self):
         )
 
 
+class Message(Base):
```
Comment (PureDreamer) [inline app/database/models.py:256]: maybe call it in another name message_notif or something?

File: app/routers/notification.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    url = router.url_path_for("view_notifications")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.get("/mark_all_as_read")
+async def mark_all_as_read(
+    db: Session = Depends(get_db)
+):
+    user_id = 1
```
Comment (PureDreamer) [inline app/routers/notification.py]: You should delete it as the sign-up is already in the site.

File: app/routers/notification.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    for m in get_all_messages(db, user_id):
```
Comment (PureDreamer) [inline app/routers/notification.py]: Please give it a better naming.

File: app/routers/notification.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    temp = notification.copy()
+    return sorted(temp, key=attrgetter('creation'), reverse=True)
+
+
+def create_message(
+        session: Session,
+        msg: str,
+        recipient_id: int,
+        link=None
+) -> None:
```
Comment (PureDreamer) [inline app/routers/notification.py]: why return none?
how should we know if it was created succesfully?

**Review by PureDreamer — COMMENTED** [review]
As always your code is clean and good!
Please look at my review

File: app/database/models.py [hunk 5]
```diff
@@ -179,6 +179,22 @@ def __repr__(self):
...
+    __tablename__ = "messages"
+
+    id = Column(Integer, primary_key=True, index=True)
+    status = Column(String, nullable=False, default="unread")
+    body = Column(String, nullable=False)
+    link = Column(String)
+    recipient_id = Column(Integer, ForeignKey("users.id"))
```
Comment (ivarshav) [inline app/database/models.py]: Move the foreign id down (for ordering)

File: app/internal/utils.py [hunk 6]
```diff
@@ -17,14 +18,22 @@ def save(session: Session, instance: Base) -> bool:
...
     save(session, instance)
     return instance
 
 
+def mark_as_read(
+        session: Session,
+        message: Union[Message, Invitation]
+) -> None:
+    """Marks a message as read."""
+    message.status = 'read'
```
Comment (ivarshav) [inline app/internal/utils.py]: It's only unread/read?
Consider Enum or boolean field

File: app/main.py [hunk 7]
```diff
@@ -54,9 +52,9 @@ def create_tables(engine, psql_environment):
     weekview.router,
     email.router,
     event.router,
+    notification.router,
```
Comment (ivarshav) [inline app/main.py]: please keep alphabetic order

File: app/routers/notification.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+router = APIRouter(
+    prefix="/notification",
+    tags=["notification"],
+    dependencies=[Depends(get_db)]
+)
+
+
+@router.get("/")
+async def view_notifications(request: Request, db: Session = Depends(get_db)):
+    user_id = 1
```
Comment (ivarshav) [inline app/routers/notification.py]: Add todo

File: app/routers/notification.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    })
+
+
+@router.post("/accept_invitations")
+async def accept_invitations(
+        request: Request,
+        db: Session = Depends(get_db)
+):
+    data = await request.form()
+    invite_id = list(data.values())[0]
```
Comment (ivarshav) [inline app/routers/notification.py]: Is it possible the list will be empty?

File: app/routers/notification.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.post("/decline_invitations")
+async def decline_invitations(
+        request: Request,
+        db: Session = Depends(get_db)
+):
+    data = await request.form()
+    invite_id = list(data.values())[0]
```
Comment (ivarshav) [inline app/routers/notification.py]: same

File: app/routers/notification.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+async def mark_all_as_read(
+    db: Session = Depends(get_db)
+):
+    user_id = 1
+    for m in get_all_messages(db, user_id):
+        if m.status == 'unread':
+            mark_as_read(db, m)
+
+    url = router.url_path_for("view_notifications")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
```
Comment (ivarshav) [inline app/routers/notification.py]: you can use starlette.status

File: app/routers/notification.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+
+async def get_message_by_id(
+        message_id: int, session: Session
+) -> Union[Message, None]:
+    """Returns a invitation by an id.
+    if id does not exist, returns None."""
+    return session.query(Message).filter_by(id=message_id).one()
```
Comment (ivarshav) [inline app/routers/notification.py]: I think that if you have no results `one` will throw an exepction.
Could you please verify that?

File: app/routers/notification.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            mark_as_read(db, m)
+
+    url = router.url_path_for("view_notifications")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+async def get_message_by_id(
+        message_id: int, session: Session
+) -> Union[Message, None]:
+    """Returns a invitation by an id.
```
Comment (ivarshav) [inline app/routers/notification.py]: an invitation

File: app/routers/notification.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+) -> List[Message]:
+    """Returns all messages."""
+    condition = Message.recipient_id == recipient_id
+    return session.query(Message).filter(condition).all()
+
+
+def get_all_invitations(session: Session, **param) -> List[Invitation]:
+    """Returns all invitations filter by param."""
+    try:
+        invitations = list(session.query(Invitation).filter_by(**param))
```
Comment (ivarshav) [inline app/routers/notification.py]: Prefer use .all() instead of list()

File: app/routers/notification.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    except SQLAlchemyError:
+        return []
+    else:
+        return invitations
+
+
+def get_invitation_by_id(
+        invitation_id: int, session: Session
+) -> Union[Invitation, None]:
+    """Returns a invitation by an id.
```
Comment (ivarshav) [inline app/routers/notification.py]: an invitation

File: app/routers/share.py [hunk 16]
```diff
@@ -62,6 +62,12 @@ def send_in_app_invitation(
     return True
 
 
+def decline(invitation: Invitation, session: Session) -> None:
+    """decline an invitation."""
+    invitation.status = 'declined'
```
Comment (ivarshav) [inline app/routers/share.py]: What are the options for invitation status? (consider Enum :))

File: app/routers/user.py [hunk 17]
```diff
@@ -7,13 +7,14 @@
 from app.internal.utils import save
 
 
-def create_user(username: str,
-                password: str,
-                email: str,
-                language_id: int,
-                session: Session) -> User:
+def create_user(
```
Comment (ivarshav) [inline app/routers/user.py]: I think it should be reverted

File: app/templates/base.html [hunk 18]
```diff
@@ -2,7 +2,7 @@
 <html lang="en">
 
 <head>
-  {% block head %}
+  {% block head %}{% endblock %}
```
Comment (ivarshav) [inline app/templates/base.html]: why?

File: app/templates/base.html [hunk 19]
```diff
@@ -14,9 +14,10 @@
     <script nomodule="" src="https://unpkg.com/ionicons@5.4.0/dist/ionicons/ionicons.js"></script>
 
     <title>PyLendar {% block title %}{% endblock %}</title>
-    {% endblock %}
 </head>
 
+
+</head>
```
Comment (ivarshav) [inline app/templates/base.html]: mistake?

File: app/templates/notification.html [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+    <div class="container mt-4">
+        <h1>{{message}}</h1>
+    </div>
+    <div id="main">
+    {% if notifications %}
+
+            <a href="{{ url_for('mark_all_as_read') }}">Mark all as read</a>
+            {% for n in notifications %}
+
+                {% if n.__table__.name == "invitations" %}
```
Comment (ivarshav) [inline app/templates/notification.html]: If possible, consider set jinja variables to reduce dups

File: tests/test_notification.py [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        assert resp.ok
+        assert self.NO_NOTIFICATIONS in resp.content
+
+    def test_accept_invitations(
+            self, user, invitation,
+            notification_test_client):
+        data = {"invite_id ": invitation.id}
+        url = router.url_path_for('accept_invitations')
+        resp = notification_test_client.post(url, data=data)
+        assert resp.status_code == HTTP_302_FOUND
```
Comment (ivarshav) [inline tests/test_notification.py]: consider use `ok`, same for all below

File: tests/test_notification.py [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+            self, user, message, sec_message,
+            notification_test_client, session
+    ):
+        url = router.url_path_for('mark_all_as_read')
+        resp = notification_test_client.get(url)
+        assert resp.status_code == HTTP_200_OK
+
+
+class TestNotification:
+
```
Comment (ivarshav) [inline tests/test_notification.py:139]: remove blank line - line 51

File: app/routers/user.py (hunk 17 above)
Comment (IdanPelled) [reply app/routers/user.py]: Why?

File: app/routers/share.py (hunk 16 above)
Comment (IdanPelled) [reply app/routers/share.py]: Great tip, thanks!

File: app/routers/notification.py (hunk 9 above)
Comment (IdanPelled) [reply app/routers/notification.py]: Nope.
except if someone sends an external post request, do you think I need to handle this scenario?

File: app/routers/notification.py (hunk 11 above)
Comment (IdanPelled) [reply app/routers/notification.py]: but I am using...

File: app/routers/notification.py (hunk 2 above)
Comment (IdanPelled) [reply app/routers/notification.py]: I talked to koby, and he said that the `current_user` feature is waiting to be merged

File: app/routers/notification.py (hunk 9 above)
Comment (ivarshav) [reply app/routers/notification.py]: at least try/except and log.error :)

File: app/routers/notification.py (hunk 11 above)
Comment (ivarshav) [reply app/routers/notification.py]: great :)

File: app/routers/user.py (hunk 17 above)
Comment (ivarshav) [reply app/routers/user.py]: We did you change the formating here?

Comment (IdanPelled) [reply app/routers/user.py]: It looked more readable to me :)

File: app/database/models.py [hunk 23]
```diff
@@ -1,11 +1,12 @@
 from __future__ import annotations
```
Comment (yammesicka) [inline app/database/models.py]: No need to put a blank line here

File: app/database/models.py [hunk 24]
```diff
@@ -172,26 +173,86 @@ class PSQLEnvironmentError(Exception):
     )
 
 
+class InvitationStatusEnum(enum.Enum):
+    unread = 0
```
Comment (yammesicka) [inline app/database/models.py]: `UNREAD`, `ACCEPTED`, `DECLINED` (Constants)

File: app/database/models.py [hunk 25]
```diff
@@ -172,26 +173,86 @@ class PSQLEnvironmentError(Exception):
...
+        participantship at the event."""
+
+        association = UserEvent(
+            user_id=self.recipient.id,
+            event_id=self.event.id
+        )
+        self.status = InvitationStatusEnum.accepted
+        session.add(self)
+        session.add(association)
+        session.commit()
```
Comment (yammesicka) [inline app/database/models.py]: What about changing from decline to accept and vice versa?

File: app/database/models.py [hunk 26]
```diff
@@ -172,26 +173,86 @@ class PSQLEnvironmentError(Exception):
...
+        nullable=False,
+    )
+
+    recipient_id = Column(Integer, ForeignKey("users.id"))
+    recipient = relationship("User")
+
+    def mark_as_read(self, session):
+        self.status = MessageStatusEnum.read
+        session.add(self)
+        session.commit()
```
Comment (yammesicka) [inline app/database/models.py:275]: Not `.merge`?

File: app/routers/invitation.py [hunk 27]
```diff
@@ -1,65 +0,0 @@
-from typing import List, Union
```
Comment (yammesicka) [inline app/routers/invitation.py]: Why did you remove it?

File: app/routers/notification.py [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+    # TODO: get current user
+    user = get_current_user(db)
+    return templates.TemplateResponse("notification.html", {
+        "request": request,
+        'new_messages': bool(get_all_messages),
+        "notifications": get_unread_notifications(session=db, user_id=user.id),
+    })
+
+
+@router.post("/accept_invitations")
```
Comment (yammesicka) [inline app/routers/notification.py]: `/invitation/accept`

File: app/routers/notification.py [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+    invite_id = list(data.values())[0]
+
+    invitation = get_invitation_by_id(invite_id, session=db)
+    invitation.accept(db)
+
+    url = router.url_path_for("view_notifications")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.post("/decline_invitations")
```
Comment (yammesicka) [inline app/routers/notification.py]: /invitation/decline

File: app/routers/notification.py [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+
+    message = await get_message_by_id(message_id, session=db)
+    if message:
+        message.mark_as_read(db)
+
+    url = router.url_path_for("view_notifications")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.get("/mark_all_as_read")
```
Comment (yammesicka) [inline app/routers/notification.py]: `/notification/read/all`?

File: app/routers/notification.py [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+    data = await request.form()
+    invite_id = list(data.values())[0]
+    invitation = get_invitation_by_id(invite_id, session=db)
+    invitation.decline(db)
+
+    url = router.url_path_for("view_notifications")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.post("/mark_as_read")
```
Comment (yammesicka) [inline app/routers/notification.py]: /notification/read/{id}

File: app/routers/notification.py [hunk 32]
```diff
@@ -0,0 +1,181 @@
...
+    user_id = 1
+    for message in get_all_messages(db, user_id):
+        if message.status == 'unread':
+            message.mark_as_read(db)
+
+    url = router.url_path_for("view_notifications")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+async def get_message_by_id(
```
Comment (yammesicka) [inline app/routers/notification.py]: From here all the way down - maybe move to internal?

File: app/routers/notification.py [hunk 33]
```diff
@@ -0,0 +1,181 @@
...
+    """Returns an invitation by an id.
+    if id does not exist, returns None."""
+    return session.query(Message).filter_by(id=message_id).first()
+
+
+def is_unread(notification):
+    return notification.status in UNREAD
+
+
+def get_unread_notifications(session: Session, user_id: int):
```
Comment (yammesicka) [inline app/routers/notification.py]: Don't forget to add type annotations to all these functions

File: app/routers/notification.py [hunk 34]
```diff
@@ -0,0 +1,181 @@
...
+
+    return sort_notifications(notifications)
+
+
+def sort_notifications(
+        notification: List[NOTIFICATION_TYPE]
+) -> List[NOTIFICATION_TYPE]:
+    """Sorts the notifications by the creation date."""
+    temp = notification.copy()
+    return sorted(temp, key=attrgetter('creation'), reverse=True)
```
Comment (yammesicka) [inline app/routers/notification.py]: Remove the temp. `sorted` already creates a copy

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great job! Added few insights :)

File: app/routers/invitation.py (hunk 27 above)
Comment (IdanPelled) [reply app/routers/invitation.py]: I didn't remove it, I converted it. Now the `notification` route contains `messages` and `invitations`

File: app/database/models.py (hunk 26 above)
Comment (IdanPelled) [reply app/database/models.py:275]: After reading a bit more, I think I should use `.flush`.

Merge:
> Used when you may have more than 1 in-memory objects which map to the same database record with some key.

Flush:
> Push changes from memory to your database's transaction buffer.

##### [Reference](https://www.michaelcho.me/article/sqlalchemy-commit-flush-expire-refresh-merge-whats-the-difference)

What do you think?

File: app/database/models.py (hunk 25 above)
Comment (IdanPelled) [reply app/database/models.py]: I agree that the user needs an option to change his mind, and I will create a `notification-history` page, where the user can change his choies.
but I still think that it will be best if the user could leve the event from the `view event` page.

File: app/database/models.py (hunk 26 above)
Comment (yammesicka) [reply app/database/models.py:275]: You want to add new record? Not to change one?

File: app/database/models.py (hunk 25 above)
Comment (yammesicka) [reply app/database/models.py]: I'm not sure I understand your point

File: app/database/models.py (hunk 26 above)
Comment (IdanPelled) [reply app/database/models.py:275]: I want to edit a record, not merge two record.
can you explain why do you think `.merge` is the best for editing a record?

File: app/database/models.py (hunk 25 above)
Comment (IdanPelled) [reply app/database/models.py]: I will try to explain my self beter.
I am now working on a page that shows all the old notifications.
In this page, the user can view old messages and accept declined invitations.
In my opinion, the user should not have an option to leave the event from this page, only from the event page.
what to you think?

Comment (yammesicka) [reply app/database/models.py]: We're currently talking about the Backend implementation.
You add a new entry for the user accepting or declining event, but technically, the current architecture won't allow me to change the response of the user from "accepted" to "declined"

File: app/database/models.py (hunk 26 above)
Comment (yammesicka) [reply app/database/models.py:275]: Flush is just making propagation for change. You changed something, and you want it to happen in the database NOW.
`merge` is taking currently existing item in the database, and either `INSERT`ing it or `UPDATE`ing it.

File: app/database/models.py [hunk 35]
```diff
@@ -201,20 +202,80 @@ class PSQLEnvironmentError(Exception):
...
+    READ = 1
+
+
 class Invitation(Base):
     __tablename__ = "invitations"
 
     id = Column(Integer, primary_key=True, index=True)
-    status = Column(String, nullable=False, default="unread")
+    creation = Column(DateTime, default=datetime.now)
+    status = Column(
```
Comment (ivarshav) [inline app/database/models.py:221]: Great!

File: app/database/models.py [hunk 36]
```diff
@@ -201,20 +202,80 @@ class PSQLEnvironmentError(Exception):
...
+    UNREAD = 0
+    READ = 1
+
+
 class Invitation(Base):
     __tablename__ = "invitations"
 
     id = Column(Integer, primary_key=True, index=True)
-    status = Column(String, nullable=False, default="unread")
+    creation = Column(DateTime, default=datetime.now)
```
Comment (ivarshav) [inline app/database/models.py]: add `nullable=False`

File: app/internal/notification.py [hunk 37]
```diff
@@ -0,0 +1,126 @@
...
+    Invitation,
+    Message,
+    InvitationStatusEnum,
+    MessageStatusEnum,
+)
+from app.internal.utils import create_model
+
+NOTIFICATION_TYPE = Union[Invitation, Message]
+
+UNREAD = [
```
Comment (ivarshav) [inline app/internal/notification.py]: consider rename to UNREAD_STATUS or similar

File: app/internal/notification.py [hunk 38]
```diff
@@ -0,0 +1,126 @@
...
+    InvitationStatusEnum.UNREAD,
+    MessageStatusEnum.UNREAD,
+]
+
+HISTORY = [
```
Comment (ivarshav) [inline app/internal/notification.py]: Maybe archived?

File: app/internal/notification.py [hunk 39]
```diff
@@ -0,0 +1,126 @@
...
+    """
+    return notification.status in HISTORY
+
+
+def get_unread_notifications(
+    session: Session,
+    user_id: int,
+) -> List[NOTIFICATION_TYPE]:
+    """Returns all unread notifications."""
+    return list(filter(_is_unread, get_all_notifications(session, user_id)))
```
Comment (ivarshav) [inline app/internal/notification.py:94]: you can pass the filter method as param and then you need only you function (and not both get_unread_notifications and get_history_notifications)

File: app/internal/notification.py [hunk 40]
```diff
@@ -0,0 +1,126 @@
...
+def get_all_notifications(
+    session: Session,
+    user_id: int,
+) -> List[NOTIFICATION_TYPE]:
+    """Returns all notifications."""
+    invitations: List[Invitation] = get_all_invitations(
+        session,
+        recipient_id=user_id,
+    )
+
```
Comment (ivarshav) [inline app/internal/notification.py]: I think this line is redundant

File: app/main.py [hunk 41]
```diff
@@ -40,10 +41,32 @@ def create_tables(engine, psql_environment):
 set_ui_language()
 
 from app.routers import (  # noqa: E402
-    about_us, agenda, calendar, categories, celebrity, credits,
-    currency, dayview, email, event, export, four_o_four, friendview,
-    google_connect, invitation, joke, login, logout, profile,
-    register, search, telegram, user, weekview, weight, whatsapp,
+    about_us,
```
Comment (ivarshav) [inline app/main.py]: Maybe revert?

File: app/main.py [hunk 42]
```diff
@@ -80,15 +103,16 @@ async def swagger_ui_redirect():
     four_o_four.router,
     friendview.router,
     google_connect.router,
-    invitation.router,
```
Comment (ivarshav) [inline app/main.py:115]: By mistake?

File: app/main.py [hunk 43]
```diff
@@ -80,15 +103,16 @@ async def swagger_ui_redirect():
...
     joke.router,
     login.router,
     logout.router,
+    notification.router,
     profile.router,
     register.router,
     salary.router,
     search.router,
     telegram.router,
+    weekview.router,
```
Comment (ivarshav) [inline app/main.py]: can be removed - see in line 117

File: app/main.py (hunk 42 above)
Comment (IdanPelled) [reply app/main.py:115]: No :)
As the description says, the `notification` route contains in it the `invitation` route

File: app/internal/notification.py (hunk 39 above)
Comment (IdanPelled) [reply app/internal/notification.py:94]: Thank's

File: app/routers/notification.py [hunk 44]
```diff
@@ -0,0 +1,174 @@
...
+    invitation = get_invitation_by_id(invite_id, session=db)
+    invitation.accept(db)
+
+    return safe_redirect_response(next)
+
+
+@router.post("/invitation/decline")
+async def decline_invitations(
+    invite_id: int = Form(...),
+    next: str = Form(...),
```
Comment (ivarshav) [inline app/routers/notification.py]: prefer not to use "next" word.

File: app/routers/register.py [hunk 45]
```diff
@@ -20,6 +20,28 @@
 )
 
 
+def _create_user(
+    username: str,
```
Comment (ivarshav) [inline app/routers/register.py]: maybe it will be easier to keep it as `**params`

File: app/templates/new_notification.html [hunk 46]
```diff
@@ -0,0 +1,108 @@
...
+                    {% if not new_messages %}disabled{% endif %} id="mark-all-ass-read">
+                        Mark all as read
+                    </button>
+                </form>
+                <div id="notifications">
+                    {% for n in notifications %}
+                        {% set type = n.__table__.name %}
+
+                        <div class="notification" id="{{ loop.index }}">
+                            {% if type == "invitations" %}
```
Comment (ivarshav) [inline app/templates/new_notification.html]: What do you think about seperate `.jinja` files for in the contents inside the if-else blocks?

File: app/main.py (hunk 41 above)
Comment (IdanPelled) [reply app/main.py]: The pre commit script changed it, I am not sure we need to revert.

File: app/internal/notification.py (hunk 39 above)
Comment (IdanPelled) [reply app/internal/notification.py:94]: I tried to implement it, but i got confused.
can you explain again what you meant?

Comment (ivarshav) [reply app/internal/notification.py:94]: You can pass the function as parameter to the function, something like:
```suggestion
    return list(filter(function_param, get_all_notifications(session, user_id)))
```

Then you don't need both get_unread_notifications and get_archived_notifications - you can keep one implementation.
Hope this helps
